bloomfilter/Data/BloomFilter/Hash.hs

```diff
@@ -165,258 +91,35 @@ evalCheapHashes (CheapHashes h1 h2) i = h1 + (h2 `unsafeShiftR` i)
-with :: Storable a => a -> (Ptr a -> IO b) -> IO b
-with val f  =
-  alloca $ \ptr -> do
-    poke ptr val
-    f ptr
-
-alignedHash2 :: Ptr a -> CSize -> Word64 -> IO Word64
-alignedHash2 ptr bytes salt =
-    with (fromIntegral salt) $ \sp -> do
-      let p1 = castPtr sp
-          p2 = castPtr sp `plusPtr` 4
-      doubleHash ptr bytes p1 p2
-      peek sp
-
-doubleHash :: Ptr a -> CSize -> Ptr Word32 -> Ptr Word32 -> IO ()
-doubleHash ptr bytes p1 p2
-          | bytes .&. 3 == 0 = hashWord2 (castPtr ptr) (div4 bytes) p1 p2
-          | otherwise        = hashLittle2 ptr bytes p1 p2
+-- | Hash 'ByteArray' given an offset and length.
+--
+-- No bounds checks are performed.
+hashByteArray :: ByteArray -> Int -> Int -> Word64 -> Word64
+hashByteArray (ByteArray ba#) off len salt =
+    unsafe_xxh3_withSeed_ba ba# (fromIntegral off) (fromIntegral len) salt
 
 instance Hashable SB.ByteString where
-    hashIO32 bs salt = unsafeUseAsCStringLen bs $ \ptr len ->
-                       alignedHash ptr (fromIntegral len) salt
-
-    {-# INLINE hashIO64 #-}
-    hashIO64 bs salt = unsafeUseAsCStringLen bs $ \ptr len ->
-                       alignedHash2 ptr (fromIntegral len) salt
-
-rechunk :: LB.ByteString -> [SB.ByteString]
-rechunk s | LB.null s = []
-          | otherwise = let (pre,suf) = LB.splitAt chunkSize s
-                        in  repack pre : rechunk suf
-    where repack    = SB.concat . LB.toChunks
-          chunkSize = fromIntegral LB.defaultChunkSize
-
-instance Hashable LB.ByteString where
-    hashIO32 bs salt = foldM (flip hashIO32) salt (rechunk bs)
-
-    {-# INLINE hashIO64 #-}
-    hashIO64 = hashChunks
-
-instance Hashable a => Hashable (Maybe a) where
-    hashIO32 Nothing salt = return salt
-    hashIO32 (Just k) salt = hashIO32 k salt
-    hashIO64 Nothing salt = return salt
-    hashIO64 (Just k) salt = hashIO64 k salt
-
-instance (Hashable a, Hashable b) => Hashable (Either a b) where
-    hashIO32 (Left a) salt = hashIO32 a salt
-    hashIO32 (Right b) salt = hashIO32 b (salt + 1)
-    hashIO64 (Left a) salt = hashIO64 a salt
-    hashIO64 (Right b) salt = hashIO64 b (salt + 1)
-
-instance (Hashable a, Hashable b) => Hashable (a, b) where
-    hashIO32 (a,b) salt = hashIO32 a salt >>= hashIO32 b
-    hashIO64 (a,b) salt = hashIO64 a salt >>= hashIO64 b
-
-instance (Hashable a, Hashable b, Hashable c) => Hashable (a, b, c) where
-    hashIO32 (a,b,c) salt = hashIO32 a salt >>= hashIO32 b >>= hashIO32 c
-
-instance (Hashable a, Hashable b, Hashable c, Hashable d) =>
-    Hashable (a, b, c, d) where
-    hashIO32 (a,b,c,d) salt =
-        hashIO32 a salt >>= hashIO32 b >>= hashIO32 c >>= hashIO32 d
-
-instance (Hashable a, Hashable b, Hashable c, Hashable d, Hashable e) =>
-    Hashable (a, b, c, d, e) where
-    hashIO32 (a,b,c,d,e) salt =
-        hashIO32 a salt >>= hashIO32 b >>= hashIO32 c >>= hashIO32 d >>= hashIO32 e
+    hashSalt64 salt bs = accursedUnutterablePerformIO $
```

Comment:
Could you fill me in on what `accursedUnutterablePerformIO` prevents here that `unsafePerformIO` wouldn't? I haven't seen this much in the wild before

Reply:
`accursedUnutterablePerformIO` is as unsafe `performIO` as it can be. But as we really don't do anything (including not allocating any temporary buffers), there is not much which can go wrong.

 `accursedUnutterablePerformIO` is what `bytestring` uses for stuff like `elemIndex` which only read the stuff, it would be unsound to use it for say `ByteString.filter` as that one would allocate new buffer, which could be alised.